#! /bin/sh -e

# gdc-driver-defaultlib.dpatch by Arthur Loiret <arthur.loiret@gmail.com>
# DP: Add -defaultlib -debuglib switches to d-spec.c.

dir=
if [ $# -eq 3 -a "$2" = '-d' ]; then
    pdir="-d $3"
    dir="$3/"
elif [ $# -ne 1 ]; then
    echo >&2 "`basename $0`: script expects -patch|-unpatch as argument"
    exit 1
fi
case "$1" in
    -patch)
        patch $pdir -f --no-backup-if-mismatch -p0 < $0
        #cd ${dir}gcc && autoconf
        ;;
    -unpatch)
        patch $pdir -f --no-backup-if-mismatch -R -p0 < $0
        #rm ${dir}gcc/configure
        ;;
    *)
        echo >&2 "`basename $0`: script expects -patch|-unpatch as argument"
        exit 1
esac
exit 0

--- gcc/d/d-spec.c	2008-02-03 20:21:18.000000000 +0000
+++ gcc/d/d-spec.c	2008-02-03 20:41:36.000000000 +0000
@@ -74,6 +74,9 @@
     /* If nonzero, the user gave us the `-v' flag.  */
     int saw_verbose_flag = 0;
 
+    /* Used by -debuglib */
+    int saw_debug_flag = 0;
+
     /* This is a tristate:
        -1 means we should not link in libstdc++
        0  means we should link in libstdc++ if it is needed
@@ -123,6 +126,9 @@
     /* The argument list.  */
     const char *const *argv;
 
+    const char* defaultlib = NULL;
+    const char* debuglib = NULL;
+
     /* The number of libraries added in.  */
     int added_libraries;
 
@@ -168,6 +174,32 @@
 			    phobos = 0;
 			    args[i] |= REMOVE_ARG;
 			}
+		    else if (strcmp (argv[i], "-defaultlib") == 0)
+		    	{
+			    added = 1;
+			    phobos = 0;
+			    args[i] |= REMOVE_ARG;
+			    if(defaultlib != NULL)
+			        free(defaultlib);
+			    defaultlib = xmalloc(sizeof(char) * (strlen(argv[++i]) + 3));
+			    strcpy(defaultlib, "-l");
+			    strcat(defaultlib, argv[i]);
+			    args[i] |= REMOVE_ARG;
+			    /* printf("XXX: arg: -defaultlib: %s\n", defaultlib); */
+			}
+		    else if (strcmp (argv[i], "-debuglib") == 0)
+		    	{
+			    added = 1;
+			    phobos = 0;
+			    args[i] |= REMOVE_ARG;
+			    if(debuglib != NULL)
+			        free(debuglib);
+			    debuglib = xmalloc(sizeof(char) * (strlen(argv[++i]) + 3));
+			    strcpy(debuglib, "-l");
+			    strcat(debuglib, argv[i]);
+			    args[i] |= REMOVE_ARG;
+			    /* printf("XXX: arg: -debuglib: %s\n", debuglib); */
+			}
 		    else if (strcmp (argv[i], "-lm") == 0
 			|| strcmp (argv[i], "-lmath") == 0
 			|| strcmp (argv[i], MATH_LIBRARY) == 0
@@ -182,6 +214,8 @@
 			saw_profile_flag++;
 		    else if (strcmp (argv[i], "-v") == 0)
 			saw_verbose_flag = 1;
+		    else if (strcmp (argv[i], "-g") == 0)
+			saw_debug_flag = 1;
 		    else if (strncmp (argv[i], "-x", 2) == 0)
 			{
 			    if (library == 0)
@@ -367,6 +401,18 @@
 	    added_libraries++;
 	    need_pthreads = 1;
 	}
+    else if (saw_debug_flag && debuglib)
+        {
+	    arglist[j++] = debuglib;
+	    added_libraries++;
+	    /* printf("XXX: debuglib: %s\n", arglist[j-1]); */
+	}
+    else if (defaultlib)
+    	{
+	    arglist[j++] = defaultlib;
+	    added_libraries++;
+	    /* printf("XXX: defaultlib: %s\n", arglist[j-1]); */
+	}
     if (saw_math)
 	arglist[j++] = saw_math;
     else if (library > 0 && need_math)
